// react-query/lib/react/index.js
"use strict";

exports.__esModule = true;
exports.useInfiniteQuery = exports.usePaginatedQuery = exports.useQuery = exports.useMutation = exports.useIsFetching = exports.useErrorResetBoundary = exports.ReactQueryErrorResetBoundary = exports.ReactQueryConfigProvider = exports.useQueryCache = exports.ReactQueryCacheProvider = void 0;

var _ReactQueryCacheProvider = require("./ReactQueryCacheProvider");

exports.ReactQueryCacheProvider = _ReactQueryCacheProvider.ReactQueryCacheProvider;
exports.useQueryCache = _ReactQueryCacheProvider.useQueryCache;

var _ReactQueryConfigProvider = require("./ReactQueryConfigProvider");

exports.ReactQueryConfigProvider = _ReactQueryConfigProvider.ReactQueryConfigProvider;

var _ReactQueryErrorResetBoundary = require("./ReactQueryErrorResetBoundary");

exports.ReactQueryErrorResetBoundary = _ReactQueryErrorResetBoundary.ReactQueryErrorResetBoundary;
exports.useErrorResetBoundary = _ReactQueryErrorResetBoundary.useErrorResetBoundary;

var _useIsFetching = require("./useIsFetching");

exports.useIsFetching = _useIsFetching.useIsFetching;

var _useMutation = require("./useMutation");

exports.useMutation = _useMutation.useMutation;

var _useQuery = require("./useQuery");

exports.useQuery = _useQuery.useQuery;

var _usePaginatedQuery = require("./usePaginatedQuery");

exports.usePaginatedQuery = _usePaginatedQuery.usePaginatedQuery;

var _useInfiniteQuery = require("./useInfiniteQuery");

exports.useInfiniteQuery = _useInfiniteQuery.useInfiniteQuery;
