// react-native-tab-view/lib/commonjs/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "TabBar", {
  enumerable: true,
  get: function get() {
    return _TabBar.default;
  }
});
Object.defineProperty(exports, "TabView", {
  enumerable: true,
  get: function get() {
    return _TabView.default;
  }
});
Object.defineProperty(exports, "TabBarIndicator", {
  enumerable: true,
  get: function get() {
    return _TabBarIndicator.default;
  }
});
Object.defineProperty(exports, "TabBarItem", {
  enumerable: true,
  get: function get() {
    return _TabBarItem.default;
  }
});
Object.defineProperty(exports, "TouchableItem", {
  enumerable: true,
  get: function get() {
    return _TouchableItem.default;
  }
});
Object.defineProperty(exports, "SceneMap", {
  enumerable: true,
  get: function get() {
    return _SceneMap.default;
  }
});
Object.defineProperty(exports, "ScrollPager", {
  enumerable: true,
  get: function get() {
    return _ScrollPager.default;
  }
});

var _TabBar = _interopRequireDefault(require("./TabBar"));

var _TabView = _interopRequireDefault(require("./TabView"));

var _TabBarIndicator = _interopRequireDefault(require("./TabBarIndicator"));

var _TabBarItem = _interopRequireDefault(require("./TabBarItem"));

var _TouchableItem = _interopRequireDefault(require("./TouchableItem"));

var _SceneMap = _interopRequireDefault(require("./SceneMap"));

var _ScrollPager = _interopRequireDefault(require("./ScrollPager"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
